README.md
# future_button
[![Coverage Status](https://coveralls.io/repos/github/kekland/future_button/badge.svg?branch=master)](https://coveralls.io/github/kekland/future_button?branch=master)
![Build status](https://img.shields.io/github/workflow/status/kekland/future_button/Test)
![Pub version](https://img.shields.io/pub/v/future_button)

Customizable Material and Cupertino buttons with progress indicators and more.

Useful if you want to run some background tasks when user presses on the button, and then show the result of the task.

![Demonstration](https://github.com/kekland/future_button/blob/master/images/future_button.gif)

## How to use it

Usage is extremely simple - right now 5 types of buttons are supported:

- `CupertinoButton` (normal + filled)
- `FlatButton` (normal + icon)
- `IconButton`
- `OutlineButton` (normal + icon)
- `RaisedButton` (normal + icon)

To convert a button to a `FutureButton`, just add a `Future` prefix and make sure that your `onPressed` callback is of type `Future<void> Function()`, instead of usual `void Function()`.

Everything that the package does can be customized. Below is the list of all parameters:

```dart
/// The widget that's used as the progress indicator.
/// You can substitute it with your own progress indicator if necessary.
/// See [defaultMaterialProgressIndicatorBuilder] and[defaultCupertinoProgressIndicatorBuilder] for more info.
final WidgetBuilder progressIndicatorBuilder;

/// The location of progress indicator.
/// See [ProgressIndicatorLocation] for more info.
final ProgressIndicatorLocation progressIndicatorLocation;

/// Whether to show the result of the Future.
/// Will show [successIndicatorBuilder] if Future completes without errors.
/// Otherwise, will show [failureIndicatorBuilder].
final bool showResult;

/// Whether to animate the transitions using [AnimatedSwitcher].
/// Defaults to [true].
final bool animateTransitions;

/// Indicator to show when the Future is completed successfully.
/// Defaults to [defaultSuccessResultIndicatorBuilder].
final WidgetBuilder successIndicatorBuilder;

/// Indicator to show when the Future fails.
/// Defaults to [defaultFailureResultIndicatorBuilder].
final WidgetBuilder failureIndicatorBuilder;

/// For how long should the result be shown for.
/// Default to `Duration(seconds: 2)`.
final Duration resultIndicatorDuration;

/// Curve that's used to animate the button size.
/// Defaults to [Curves.easeInOut].
final Curve animationCurve;

/// Duration that's used to animate the button size.
/// Defaults to `Duration(milliseconds: 150)`.
final Duration animationDuration;
```

## Example

```dart
FutureRaisedButton.icon(
  icon: Icon(Icons.star),
  label: Text('Star'),
  onPressed: () async {
    // Do some background task
    await Future.delayed(Duration(seconds: 5));
  },
);
```

See `example` project for more details.

## Getting Started

### Depend on it

Add this to your package's pubspec.yaml file:

```
dependencies:
  future_button: ^0.1.1
```

### Install it

You can install packages from the command line:

```bash
$ flutter pub get
```

### Import it

Now in your Dart code, you can use:

```dart
import 'package:future_button/future_button.dart';
```

## Contact me

**E-mail**: `kk.erzhan@gmail.com`

CHANGELOG.md
## [0.1.2] - 04.10.2020

* Fix missing icons from Flutter v1.22.0

## [0.1.1] - 04.10.2020

* Update `README.md`

## [0.1.0] - 04.10.2020

* Support for showing the result of the action after the Future is completed

## [0.0.2] - 04.10.2020

* Support for exceptions thrown in `onPressed` callback
* Documentation improvements

## [0.0.1] - 01.10.2020

* Initial release


test/future_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:future_button/future_button.dart';

class PlannedException implements Exception {}

const waitDuration = Duration(milliseconds: 150);
const animationDuration = Duration(milliseconds: 50);
const resultIndicatorDuration = Duration(milliseconds: 150);

final progressIndicatorBuilders = <WidgetBuilder>[
  defaultMaterialProgressIndicatorBuilder,
  defaultCupertinoProgressIndicatorBuilder,
];

final successIndicatorBuilder = defaultSuccessResultIndicatorBuilder;
final failureIndicatorBuilder = defaultFailureResultIndicatorBuilder;

typedef FutureButtonBuilder = Widget Function({
  FutureCallback onPressed,
  WidgetBuilder progressIndicatorBuilder,
  WidgetBuilder successIndicatorBuilder,
  WidgetBuilder failureIndicatorBuilder,
  bool showResult,
  Widget child,
  ProgressIndicatorLocation progressIndicatorLocation,
});

Future<void> waitForAndFail() async {
  await Future.delayed(waitDuration);
  throw PlannedException();
}

Future<void> waitFor() {
  return Future.delayed(waitDuration);
}

Future<void> waitForAnimation() {
  return Future.delayed(animationDuration);
}

Future<void> waitForResultIndicator() {
  return Future.delayed(resultIndicatorDuration);
}

Future<void> testButtonWithArgs(
  WidgetTester tester, {
  List<ProgressIndicatorLocation> progressIndicatorLocations =
      ProgressIndicatorLocation.values,
  FutureButtonBuilder builder,
  FutureCallback onTap,
  bool shouldError = false,
  bool shouldShowResultIndicator = false,
}) async {
  for (final progressIndicatorLocation in progressIndicatorLocations) {
    for (final progressIndicatorBuilder in progressIndicatorBuilders) {
      final child = Container();
      final progressIndicator = progressIndicatorBuilder(null);
      final successIndicator = progressIndicatorBuilder(null);
      final failureIndicator = progressIndicatorBuilder(null);

      final widget = builder(
        onPressed: onTap,
        progressIndicatorBuilder: (_) => progressIndicator,
        successIndicatorBuilder: (_) => successIndicator,
        failureIndicatorBuilder: (_) => failureIndicator,
        progressIndicatorLocation: progressIndicatorLocation,
        showResult: shouldShowResultIndicator,
        child: child,
      );

      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(body: Center(child: widget)),
        ),
      );

      final state = tester.state<GenericFutureButtonState>(
        find.byWidget(widget),
      );

      expect(find.byWidget(widget), findsOneWidget);
      expect(find.byWidget(child), findsOneWidget);
      expect(find.byWidget(progressIndicator), findsNothing);

      expect(state.isLoading, equals(false));
      expect(state.isEnabled, equals(onTap != null));

      if (onTap != null) {
        await tester.runAsync(() async {
          await tester.tap(find.byWidget(widget));
          await tester.pump();

          expect(
            find.byWidget(child),
            progressIndicatorLocation == ProgressIndicatorLocation.center
                ? findsNothing
                : findsOneWidget,
          );

          expect(find.byWidget(progressIndicator), findsOneWidget);

          expect(state.isLoading, equals(true));
          expect(state.isEnabled, equals(false));

          await waitFor();
          await tester.pump();
        });

        if (shouldShowResultIndicator) {
          expect(find.byWidget(widget), findsOneWidget);

          if (progressIndicatorLocation == ProgressIndicatorLocation.center) {
            expect(find.byWidget(child), findsNothing);
          } else {
            expect(find.byWidget(child), findsOneWidget);
          }

          expect(find.byWidget(progressIndicator), findsNothing);
          expect(
              find.byWidget(shouldError ? failureIndicator : successIndicator),
              findsOneWidget);

          expect(state.isLoading, equals(true));
          expect(state.isEnabled, equals(false));

          await tester.runAsync(() async {
            await waitForResultIndicator();
            await tester.pump();
          });

          expect(find.byWidget(widget), findsOneWidget);
          expect(find.byWidget(child), findsOneWidget);
          expect(find.byWidget(successIndicator), findsNothing);
          expect(find.byWidget(failureIndicator), findsNothing);
          expect(find.byWidget(progressIndicator), findsNothing);

          expect(state.isLoading, equals(false));
          expect(state.isEnabled, equals(true));
        } else {
          expect(find.byWidget(widget), findsOneWidget);
          expect(find.byWidget(child), findsOneWidget);
          expect(find.byWidget(progressIndicator), findsNothing);

          expect(state.isLoading, equals(false));
          expect(state.isEnabled, equals(true));
        }
      }
    }
  }
}

void testButton({
  String name,
  FutureButtonBuilder builder,
  List<ProgressIndicatorLocation> progressIndicatorLocations =
      ProgressIndicatorLocation.values,
}) {
  testWidgets(
    'Test normal $name',
    (tester) async {
      await testButtonWithArgs(
        tester,
        builder: builder,
        onTap: waitFor,
        progressIndicatorLocations: progressIndicatorLocations,
      );
    },
  );

  testWidgets(
    'Test disabled $name',
    (tester) async {
      await testButtonWithArgs(
        tester,
        builder: builder,
        onTap: null,
        progressIndicatorLocations: progressIndicatorLocations,
      );
    },
  );

  testWidgets(
    'Test success result $name',
    (tester) async {
      await testButtonWithArgs(
        tester,
        builder: builder,
        onTap: waitFor,
        progressIndicatorLocations: progressIndicatorLocations,
        shouldError: false,
        shouldShowResultIndicator: true,
      );
    },
  );

  testWidgets(
    'Test failure result $name',
    (tester) async {
      await testButtonWithArgs(
        tester,
        builder: builder,
        onTap: waitForAndFail,
        progressIndicatorLocations: progressIndicatorLocations,
        shouldError: true,
        shouldShowResultIndicator: true,
      );
    },
  );
}

void main() {
  testButton(
    name: 'FutureCupertinoButton',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureCupertinoButton(
        child: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureCupertinoButton.filled',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureCupertinoButton.filled(
        child: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureFlatButton',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureFlatButton(
        child: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureFlatButton.icon',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureFlatButton.icon(
        icon: Icon(Icons.star),
        label: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );
  testButton(
    name: 'FutureIconButton',
    progressIndicatorLocations: [ProgressIndicatorLocation.center],
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureIconButton(
        icon: child,
        onPressed: onPressed,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureOutlineButton',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureOutlineButton(
        child: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureOutlineButton.icon',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureOutlineButton.icon(
        icon: Icon(Icons.star),
        label: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureRaisedButton',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureRaisedButton(
        child: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );

  testButton(
    name: 'FutureRaisedButton.icon',
    builder: ({
      FutureCallback onPressed,
      WidgetBuilder progressIndicatorBuilder,
      WidgetBuilder successIndicatorBuilder,
      WidgetBuilder failureIndicatorBuilder,
      bool showResult,
      Widget child,
      ProgressIndicatorLocation progressIndicatorLocation,
    }) {
      return FutureRaisedButton.icon(
        icon: Icon(Icons.star),
        label: child,
        onPressed: onPressed,
        progressIndicatorLocation: progressIndicatorLocation,
        progressIndicatorBuilder: progressIndicatorBuilder,
        successIndicatorBuilder: successIndicatorBuilder,
        failureIndicatorBuilder: failureIndicatorBuilder,
        showResult: showResult,
        animationDuration: animationDuration,
        animateTransitions: false,
        resultIndicatorDuration: resultIndicatorDuration,
      );
    },
  );
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:future_button/future_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Future Button'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  MyHomePage({
    Key key,
    this.title,
  }) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Future<void> waitFor([int seconds = 2]) {
    return Future.delayed(Duration(seconds: seconds));
  }

  Future<void> waitForAndFail([int seconds = 2]) async {
    await Future.delayed(Duration(seconds: seconds));
    throw Exception();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FutureRaisedButton(
              child: Text('Load'),
              onPressed: waitFor,
              showResult: true,
            ),
            FutureRaisedButton(
              child: Text('Load (right)'),
              onPressed: waitForAndFail,
              progressIndicatorLocation: ProgressIndicatorLocation.right,
              showResult: true,
            ),
            FutureRaisedButton.icon(
              icon: Icon(Icons.star),
              label: Text('Star'),
              onPressed: waitFor,
              showResult: false,
            ),
            FutureIconButton(
              icon: Icon(Icons.link),
              onPressed: waitFor,
              showResult: true,
            ),
            FutureCupertinoButton(
              child: Text('Like'),
              onPressed: waitFor,
              showResult: true,
            ),
            FutureCupertinoButton.filled(
              child: Text('Like'),
              onPressed: waitForAndFail,
              showResult: true,
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/buttons/material/future_icon_button.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:future_button/src/generic_future_button.dart';

/// A Future version of the material [IconButton].
/// See [IconButton] for parameter descriptions.
class FutureIconButton extends GenericFutureButtonWidget {
  final double iconSize;
  final VisualDensity visualDensity;
  final EdgeInsetsGeometry padding;
  final AlignmentGeometry alignment;
  final double splashRadius;
  final Color focusColor;
  final Color hoverColor;
  final Color color;
  final Color splashColor;
  final Color highlightColor;
  final Color disabledColor;
  final MouseCursor mouseCursor;
  final FocusNode focusNode;
  final bool autofocus;
  final String tooltip;
  final bool enableFeedback;
  final BoxConstraints constraints;

  /// Default [FutureIconButton] constructor.
  /// See [IconButton] for more info.
  FutureIconButton({
    Key key,
    @required FutureCallback onPressed,
    @required Widget icon,
    Curve animationCurve,
    Duration animationDuration,
    bool showResult = false,
    bool animateTransitions = true,
    WidgetBuilder progressIndicatorBuilder,
    WidgetBuilder successIndicatorBuilder,
    WidgetBuilder failureIndicatorBuilder,
    Duration resultIndicatorDuration,
    this.iconSize = 24.0,
    this.visualDensity,
    this.padding = const EdgeInsets.all(8.0),
    this.alignment = Alignment.center,
    this.splashRadius,
    this.color,
    this.focusColor,
    this.hoverColor,
    this.highlightColor,
    this.splashColor,
    this.disabledColor,
    this.mouseCursor = SystemMouseCursors.click,
    this.focusNode,
    this.autofocus = false,
    this.tooltip,
    this.enableFeedback = true,
    this.constraints,
  }) : super(
          key: key,
          child: icon,
          onPressed: onPressed,
          animationCurve: animationCurve,
          animationDuration: animationDuration,
          showResult: showResult,
          animateTransitions: animateTransitions,
          progressIndicatorBuilder: progressIndicatorBuilder,
          successIndicatorBuilder: successIndicatorBuilder,
          failureIndicatorBuilder: failureIndicatorBuilder,
          resultIndicatorDuration: resultIndicatorDuration,
          progressIndicatorLocation: ProgressIndicatorLocation.center,
        );

  @override
  _FutureIconButtonState createState() => _FutureIconButtonState();
}

class _FutureIconButtonState
    extends GenericFutureButtonState<FutureIconButton> {
  @override
  Widget buildButton({
    BuildContext context,
    Widget child,
    VoidCallback onPressed,
  }) {
    return IconButton(
      icon: child,
      onPressed: isEnabled ? onPressed : null,
      iconSize: widget.iconSize,
      visualDensity: widget.visualDensity,
      padding: widget.padding,
      alignment: widget.alignment,
      splashRadius: widget.splashRadius,
      color: widget.color,
      focusColor: widget.focusColor,
      hoverColor: widget.hoverColor,
      highlightColor: widget.highlightColor,
      splashColor: widget.splashColor,
      disabledColor: widget.disabledColor,
      mouseCursor: widget.mouseCursor,
      focusNode: widget.focusNode,
      autofocus: widget.autofocus,
      tooltip: widget.tooltip,
      enableFeedback: widget.enableFeedback,
      constraints: widget.constraints,
    );
  }
}


lib/src/generic_future_button.dart
import 'package:flutter/widgets.dart';
import 'package:flutter/material.dart' as material;
import 'package:flutter/cupertino.dart' as cupertino;

/// An `onPressed` function type for [GenericFutureButtonWidget].
/// It's different from default [VoidCallback] in a sense that the action of the widget is asynchronous.
typedef FutureCallback = Future<void> Function();

/// A default Material progress indicator. It uses [Material.CircularProgressIndicator].
/// It's used by default in [FutureFlatButton], [FutureIconButton], [FutureOutlineButton] and [FutureRaisedButton].
final WidgetBuilder defaultMaterialProgressIndicatorBuilder =
    (context) => SizedBox(
          width: 20.0,
          height: 20.0,
          child: material.CircularProgressIndicator(strokeWidth: 2.0),
        );

/// A default Cupertino progress indicator. It uses [Cupertino.CupertinoActivityIndicator].
/// It's used by default in [FutureCupertinoButton].
final WidgetBuilder defaultCupertinoProgressIndicatorBuilder =
    (context) => SizedBox(
          width: 20.0,
          height: 20.0,
          child: cupertino.CupertinoActivityIndicator(radius: 8.0),
        );

/// A default widget for "success" result indicator.
final WidgetBuilder defaultSuccessResultIndicatorBuilder =
    (context) => Container(
          width: 20.0,
          height: 20.0,
          decoration: BoxDecoration(
            color: material.Colors.green,
            shape: BoxShape.circle,
          ),
          alignment: Alignment.center,
          child: Icon(
            material.Icons.check,
            size: 16.0,
            color: material.Colors.white,
          ),
        );

/// A default widget for "failed" result indicator.
final WidgetBuilder defaultFailureResultIndicatorBuilder =
    (context) => Container(
          width: 20.0,
          height: 20.0,
          decoration: BoxDecoration(
            color: material.Colors.red,
            shape: BoxShape.circle,
          ),
          alignment: Alignment.center,
          child: Icon(
            material.Icons.error_outline,
            size: 16.0,
            color: material.Colors.white,
          ),
        );

/// The location of the progress indicator.
/// [left] and [right] means that the progress indicator will be on the left or on the right, respectively.
/// [center] means that the progress indicator will be centered and will hide [child].
enum ProgressIndicatorLocation {
  left,
  center,
  right,
}

enum FutureButtonState {
  success,
  failed,
  progress,
  normal,
}

/// A generic class for all Future buttons.
/// See [FutureRaisedButton] or any FutureButton for info on how to create your own Future button.
abstract class GenericFutureButtonWidget extends StatefulWidget {
  /// A callback of type `Future<void> Function()`. This will be ran when user presses on the button.
  /// If the future is in progress, the loading indicator will be displayed.
  /// When the future completes (regardless if successful or not), the loading indicator will dissapear.
  final FutureCallback onPressed;

  /// The widget that's used as the progress indicator.
  /// You can substitute it with your own progress indicator if necessary.
  /// See [defaultMaterialProgressIndicatorBuilder] and [defaultCupertinoProgressIndicatorBuilder] for more info.
  final WidgetBuilder progressIndicatorBuilder;

  /// The child of the widget. Usually a [Text] widget.
  /// It will be hidden if [progressIndicatorLocation] is [ProgressIndicatorLocation.center] and future is in progress.
  final Widget child;

  /// The location of progress indicator.
  /// See [ProgressIndicatorLocation] for more info.
  final ProgressIndicatorLocation progressIndicatorLocation;

  /// Whether to show the result of the Future.
  /// Will show [successIndicatorBuilder] if Future completes without errors.
  /// Otherwise, will show [failureIndicatorBuilder].
  final bool showResult;

  /// Whether to animate the transitions using [AnimatedSwitcher].
  /// Defaults to [true].
  final bool animateTransitions;

  /// Indicator to show when the Future is completed successfully.
  /// Defaults to [defaultSuccessResultIndicatorBuilder].
  final WidgetBuilder successIndicatorBuilder;

  /// Indicator to show when the Future fails.
  /// Defaults to [defaultFailureResultIndicatorBuilder].
  final WidgetBuilder failureIndicatorBuilder;

  /// For how long should the result be shown for.
  /// Default to `Duration(seconds: 2)`.
  final Duration resultIndicatorDuration;

  /// Curve that's used to animate the button size.
  /// Defaults to [Curves.easeInOut].
  final Curve animationCurve;

  /// Duration that's used to animate the button size.
  /// Defaults to `Duration(milliseconds: 150)`.
  final Duration animationDuration;

  const GenericFutureButtonWidget({
    Key key,
    @required this.onPressed,
    @required this.child,
    this.progressIndicatorBuilder,
    this.successIndicatorBuilder,
    this.failureIndicatorBuilder,
    this.showResult = false,
    this.animateTransitions = true,
    ProgressIndicatorLocation progressIndicatorLocation,
    Curve animationCurve,
    Duration animationDuration,
    Duration resultIndicatorDuration,
  })  : progressIndicatorLocation =
            progressIndicatorLocation ?? ProgressIndicatorLocation.left,
        animationDuration =
            animationDuration ?? const Duration(milliseconds: 150),
        animationCurve = animationCurve ?? Curves.easeInOut,
        resultIndicatorDuration =
            resultIndicatorDuration ?? const Duration(seconds: 2),
        super(key: key);
}

/// This is the state of the [GenericFutureButtonWidget].
/// If you want to implement your own Future button, take a look at [FutureRaisedButton] or other buttons.
abstract class GenericFutureButtonState<T extends GenericFutureButtonWidget>
    extends State<T> with SingleTickerProviderStateMixin {
  FutureButtonState _state = FutureButtonState.normal;

  /// Whether the future is in progress.
  bool get isLoading => _state != FutureButtonState.normal;

  /// Whether the button can be pressed.
  bool get isEnabled => !isLoading && widget.onPressed != null;

  /// The builder for the indicator.
  /// Returned widget is dependent on [_state] and can be one of
  /// [_successIndicatorBuilder], [_failureIndicatorBuilder] or [_progressIndicatorBuilder].
  WidgetBuilder get _indicatorBuilder {
    switch (_state) {
      case FutureButtonState.success:
        return _successIndicatorBuilder;
      case FutureButtonState.failed:
        return _failureIndicatorBuilder;
      case FutureButtonState.progress:
        return _progressIndicatorBuilder;
      default:
        return null;
    }
  }

  /// The builder for the progress indicator.
  /// If none is passed, uses [defaultMaterialProgressIndicatorBuilder].
  WidgetBuilder get _progressIndicatorBuilder =>
      widget.progressIndicatorBuilder ??
      defaultMaterialProgressIndicatorBuilder;

  /// The builder for the success indicator.
  /// If none is passed, uses [defaultSuccessIndicatorBuilder].
  WidgetBuilder get _successIndicatorBuilder =>
      widget.successIndicatorBuilder ?? defaultSuccessResultIndicatorBuilder;

  /// The builder for the failure indicator.
  /// If none is passed, uses [defaultFailureIndicatorBuilder].
  WidgetBuilder get _failureIndicatorBuilder =>
      widget.failureIndicatorBuilder ?? defaultFailureResultIndicatorBuilder;

  /// [TextDirection] is used to indicate the order of the widgets in the button.
  /// If [widget.progressIndicatorLocation] is [ProgressIndicatorLocation.left], the order is normal.
  /// Otherwise, the order is reversed.
  TextDirection get textDirection {
    switch (widget.progressIndicatorLocation) {
      case ProgressIndicatorLocation.left:
        return TextDirection.ltr;
      case ProgressIndicatorLocation.right:
        return TextDirection.rtl;
      default:
        return null;
    }
  }

  /// Aligning the animation correctly is important.
  /// Here the animation must be aligned to the opposite side of the progress indicator so that the text doesn't get cut off.
  Alignment get animationAlignment {
    switch (widget.progressIndicatorLocation) {
      case ProgressIndicatorLocation.left:
        return Alignment.centerRight;
      case ProgressIndicatorLocation.center:
        return Alignment.center;
      case ProgressIndicatorLocation.right:
        return Alignment.centerLeft;
      default:
        return null;
    }
  }

  /// The body of the button.
  /// If the future is completed or not running, the body is [widget.child].
  /// If the future is running and [widget.progressIndicatorLocation] is [ProgressIndicatorLocation.center], it'll display the progress indicator only.
  /// If the future is running and [widget.progressIndicatorLocation] is not [ProgressIndicatorLocation.center], it'll display the progress indicator and the child. The order depends on [textDirection].
  Widget get child {
    Widget child;

    if (!isLoading) {
      child = widget.child;
    } else if (widget.progressIndicatorLocation ==
        ProgressIndicatorLocation.center) {
      child = _indicatorBuilder(context);
    } else {
      return Row(
        mainAxisSize: MainAxisSize.min,
        textDirection: textDirection,
        children: [
          AnimatedSwitcher(
            duration: widget.animationDuration,
            switchInCurve: widget.animationCurve,
            switchOutCurve: widget.animationCurve,
            child: _indicatorBuilder(context),
          ),
          SizedBox(width: 8.0),
          widget.child,
        ],
      );
    }

    if (!widget.animateTransitions) return child;

    return AnimatedSwitcher(
      duration: widget.animationDuration,
      switchInCurve: widget.animationCurve,
      switchOutCurve: widget.animationCurve,
      child: child,
    );
  }

  /// The `onPressed` callback wrapper.
  /// It awaits for the future to be finished.
  Future<void> onPressed() async {
    setState(() {
      _state = FutureButtonState.progress;
    });

    Exception error;
    try {
      await widget.onPressed();
    } catch (e) {
      error = e;
    }

    if (!widget.showResult) {
      setState(() {
        _state = FutureButtonState.normal;
      });

      if (error != null) {
        throw error;
      }
    } else {
      setState(() {
        _state = error == null
            ? FutureButtonState.success
            : FutureButtonState.failed;
      });

      await Future.delayed(widget.resultIndicatorDuration, () {
        setState(() {
          _state = FutureButtonState.normal;
        });
      });
    }
  }

  /// This is an abstract function that's dependent on the button type.
  /// See [FutureRaisedButton] or other buttons for implementation details.
  Widget buildButton({
    BuildContext context,
    Widget child,
    VoidCallback onPressed,
  });

  @override
  Widget build(BuildContext context) {
    return buildButton(
      context: context,
      child: AnimatedSize(
        vsync: this,
        duration: widget.animationDuration,
        curve: widget.animationCurve,
        alignment: animationAlignment,
        child: child,
      ),
      onPressed: isEnabled ? onPressed : null,
    );
  }
}
